connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -307,15 +307,25 @@ private void processConnectorConfigUpdates(Set<String> connectorConfigUpdates) {
     }
 
     private void processTargetStateChanges(Set<String> connectorTargetStateChanges) {
-        if (!connectorTargetStateChanges.isEmpty()) {
-            for (String connector : connectorTargetStateChanges) {
-                if (worker.connectorNames().contains(connector)) {
-                    TargetState targetState = configState.targetState(connector);
-                    worker.setTargetState(connector, targetState);
-                    if (targetState == TargetState.STARTED)
-                        reconfigureConnectorTasksWithRetry(connector);
-                }
+        for (String connector : connectorTargetStateChanges) {
+            TargetState targetState = configState.targetState(connector);
+            if (targetState == null) {
+                // The target state was deleted. This should follow only after deletion of the
+                // connector's configuration itself, so raise an error if otherwise.
+                if (configState.connectors().contains(connector))
+                    throw new ConnectException("Unexpected removal of target state for connector " + connector);
```

Comment:
Even if we write it before we still need to deal with the compatibility issue during upgrades, right? So we could start writing it now, but until we went through at least one more upgrade, I think we'd still only be able to log a message for the time being.